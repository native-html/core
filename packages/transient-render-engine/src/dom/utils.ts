import { Text as TextType, Tag as TagType } from 'domelementtype';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === TextType;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === TagType;
}
