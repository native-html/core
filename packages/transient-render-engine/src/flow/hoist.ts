import TBlockCtor, { TBlockImpl } from '../tree/TBlockCtor';
import { TNodeImpl } from '../tree/tree-types';
import { TPhrasingCtor } from '../tree/TPhrasingCtor';
import { TTextCtor } from '../tree/TTextCtor';

/**
 * Wrap text nodes around TPhrasing nodes.
 * @param tnode The parent node of all elements to group.
 */
function groupText(tnode: TBlockImpl): TNodeImpl {
  let newChildren: TNodeImpl[] = [];
  const wrapperInit = {
    elementModel: null,
    context: tnode.init.context,
    // We need to merge styles here to make sure
    // some React Native styles working only for the uppermost Text element
    // such as "textAlign" are preserved.
    parentStyles: tnode.styles,
    parent: null
  };
  let wrapper = new TPhrasingCtor(wrapperInit);
  let wrapperChildren: TNodeImpl[] = [];
  for (const child of tnode.children) {
    if (child instanceof TTextCtor || child instanceof TPhrasingCtor) {
      wrapperChildren.push(child);
    } else {
      if (wrapperChildren.length) {
        newChildren.push(wrapper);
        wrapper.bindChildren(wrapperChildren);
        wrapper = new TPhrasingCtor(wrapperInit);
        wrapperChildren = [];
      }
      newChildren.push(child);
    }
  }
  if (wrapperChildren.length) {
    wrapper.bindChildren(wrapperChildren);
    newChildren.push(wrapper);
  }
  tnode.bindChildren(newChildren);
  return tnode;
}

function hoistNode(tnode: TNodeImpl): TNodeImpl {
  if (tnode instanceof TTextCtor) {
    return tnode;
  }
  tnode.bindChildren(tnode.children.map(hoistNode));
  if (tnode instanceof TPhrasingCtor) {
    for (const cnode of tnode.children) {
      if (cnode instanceof TBlockCtor) {
        const newNode = new TBlockCtor(tnode.cloneInitParams());
        newNode.bindChildren(tnode.children);
        const output = groupText(newNode);
        return output;
      }
    }
  } else if (tnode instanceof TBlockCtor) {
    if (tnode.children.length > 0) {
      const output = groupText(tnode);
      return output;
    }
  }
  return tnode;
}

export function hoist(tree: TNodeImpl): TNodeImpl {
  return hoistNode(tree);
}
