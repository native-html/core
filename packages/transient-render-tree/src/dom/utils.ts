import { ElementType } from 'htmlparser2';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === ElementType.Text;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === ElementType.Tag;
}
