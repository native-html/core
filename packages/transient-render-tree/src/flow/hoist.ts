import { TBlock } from '../tree/TBlock';
import { TBlockAnchor } from '../tree/TBlockAnchor';
import { TNode } from '../tree/TNode';
import { TPhrasing } from '../tree/TPhrasing';
import { TPhrasingAnchor } from '../tree/TPhrasingAnchor';
import { TText } from '../tree/TText';

function groupText(tnode: TBlock, wrappernode: TPhrasing): TNode {
  let newChildren: TNode[] = [];
  let wrapper = wrappernode.newEmpty();
  for (const child of tnode.children) {
    const newChild = child;
    if (newChild instanceof TText || newChild instanceof TPhrasing) {
      wrapper.children.push(newChild);
    } else {
      newChildren.push(wrapper);
      wrapper = wrappernode.newEmpty();
      if (wrappernode instanceof TPhrasingAnchor) {
        const nextChild = new TBlockAnchor({
          ...newChild.cloneInitParams(),
          href: wrappernode.href,
          parentStyles: wrappernode.parentStyles
        });
        newChildren.push(nextChild);
      } else {
        newChildren.push(newChild);
      }
    }
  }
  if (wrapper.children.length) {
    newChildren.push(wrapper);
  }
  tnode.bindChildren(newChildren);
  return tnode;
}

function hoistNode(node: TNode): TNode {
  node.bindChildren(node.children.map(hoistNode));
  if (node instanceof TPhrasing) {
    for (const child of node.children) {
      if (child instanceof TBlock) {
        const newNode = new TBlock(
          node.cloneInitParams({
            parentStyles: child.parentStyles,
            styles: child.styles
          })
        );
        newNode.bindChildren(node.children);
        return groupText(newNode, node);
      }
    }
  } else if (node instanceof TBlock) {
    if (node.children.length > 0) {
      return groupText(
        node,
        new TPhrasing({
          parentStyles: node.styles,
          stylesMerger: node.stylesMerger
        })
      );
    }
  }
  return node;
}

export function hoist(tree: TNode): TNode {
  return hoistNode(tree);
}
