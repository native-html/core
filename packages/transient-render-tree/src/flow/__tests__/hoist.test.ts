import { imgSrc, rfc002Source, href } from './shared';
import { hoist } from '../hoist';
import { translateTreeTest } from './utils';

function makeTTree(source: string) {
  return hoist(translateTreeTest(source));
}

describe('hoist function', () => {
  it('should comply with RFC002 example (hoisting) ', () => {
    const hoistedTree = makeTTree(rfc002Source);
    expect(hoistedTree).toMatchObject({
      type: 'block',
      isAnchor: false,
      tagName: 'a',
      attributes: { href }, // Should preserve attributes in TBlock
      children: [
        {
          type: 'phrasing',
          isAnchor: true,
          tagName: null,
          href,
          attributes: {}
        },
        {
          type: 'block',
          isAnchor: true,
          tagName: 'img',
          href,
          attributes: {
            src: imgSrc
          }
        },
        {
          type: 'phrasing',
          isAnchor: true,
          tagName: null,
          attributes: {},
          href
        }
      ]
    });
    expect(hoistedTree).toMatchSnapshot();
  });
  it('should preserve styles of encompassing blocks', () => {
    const tdoc = makeTTree(
      '<a href="http://google.fr"><div style="background-color: red;"></div></a>'
    );
    const tdiv = tdoc.children[0];
    expect(tdiv).toMatchObject({
      type: 'block',
      tagName: 'div'
    });
    expect(tdoc.children).toHaveLength(1);
    expect(tdiv.styles.nativeBlockRet.backgroundColor).toBe('red');
  });
});
