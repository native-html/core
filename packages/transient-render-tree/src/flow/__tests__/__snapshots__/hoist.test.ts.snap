// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoist function should comply with RFC002 example (hoisting)  1`] = `
TBlock {
  "attributes": Object {
    "href": "https://domain.com",
  },
  "children": Array [
    TPhrasingAnchor {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": "
This is
",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": "phrasing content",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "span",
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": "
",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
      ],
      "className": null,
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "id": null,
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
    TBlockAnchor {
      "attributes": Object {
        "src": "https://domain.com/logo.jpg",
      },
      "children": Array [],
      "className": null,
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "id": null,
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "img",
      "type": "block",
    },
    TPhrasingAnchor {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": "
    and this is ",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": "too",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "strong",
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "className": null,
          "data": ".
",
          "hasWhiteSpaceCollapsingEnabled": true,
          "id": null,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
      ],
      "className": null,
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "id": null,
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
  ],
  "className": null,
  "hasWhiteSpaceCollapsingEnabled": true,
  "id": null,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "a",
  "type": "block",
}
`;
